login: Add show password feature.

ligmitz opened the pull request:
Fixes #17301

<!-- What's this PR for?  (Just a link to an issue is fine.) -->


**Testing plan:** <!-- How have you tested? -->
Dev server

**GIFs or screenshots:** <!-- If a UI change.  See:
  https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
  -->
![eye-icon](https://user-images.githubusercontent.com/55033316/107934849-93365d80-6fa6-11eb-959f-e4d4e8a17ef3.gif)


<!-- Also be sure to make clear, coherent commits:
  https://zulip.readthedocs.io/en/latest/contributing/version-control.html
  -->

showell commented:
This seems like the right approach to me, and the code is clean.

The eye icon feels intuitive to me.

It's a shame that there's not a more native-to-browser way to implement this (*), at least as I understand it.  When you click on the eye icon in this PR, we change the field from `type="password"` to `type="text"`, which means that the browser can't remember your password for you.  And it's particularly annoying that this will impact folks who have issues typing their passwords.

But this seems like a no-brainer to merge.

* See https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/password

ligmitz commented:
@showell I have changed the default icon as `eye` instead of `eye-slash` for the i tag considering the convention. Regarding the password saving feature supported by browser, I have added the `autocomplete: current-password` which uses the username-password saved for the site as default credentials, which I think will solve the mentioned problem.

timabbott commented:
@ligmitz can you post screenshots of the similar feature in a couple popular sites that are thoughtful about security (E.g. Dropbox, )?  And maybe also look to see if they're using the same HTML password<=>text field strategy, or something different.  I'd like to make sure we're using best practices for this sort of widget in terms of both appearance and mechanism.

I'd like for @andersk to do a security review of this before we merged as well; we want to make sure we're thinking through any change around passwords.

I guess we should also look at how password managers like Lastpass (that insert their own elements at the right end of password boxes) do with this.

timabbott commented:
A few things worth looking at:

* https://security.stackexchange.com/questions/185851/pitfalls-when-rolling-your-own-show-password-functionality
* Several other things one finds searching for "show password feature security" on Google.

@ligmitz perhaps you can do a bit of research, reading those and posting your findings?

ligmitz commented:
> @ligmitz can you post screenshots of the similar feature in a couple popular sites that are thoughtful about security (E.g. Dropbox, )? And maybe also look to see if they're using the same HTML password<=>text field strategy, or something different. I'd like to make sure we're using best practices for this sort of widget in terms of both appearance and mechanism.

@timabbott I went through the following sites as you asked:
* Google
Google uses the same approach for showing the password i.e. the inter-conversion between `password`<=>`text` . One of the additional thing they perform is changing the `autocomplete` attr from `current-password` to `off` on displaying the password.

With text - 
![gmail1](https://user-images.githubusercontent.com/55033316/108069590-8ccce200-7089-11eb-8cd8-089fc163c063.png)

With password -
![gmail2](https://user-images.githubusercontent.com/55033316/108069667-a110df00-7089-11eb-8a5e-0cc8f8feef57.png)

* Dropbox
Dropbox does not have the feature to display the password. However they set the `value` inside the password `input` tag with the value of the field which is kind of insecure as it exposes the password directly to the browser. One of the things to notice is the value of the `autocomplete` attr is set as `off`.

![dropbox](https://user-images.githubusercontent.com/55033316/108070126-25fbf880-708a-11eb-9c28-34ca8a69859e.png)

* lastpass
While going through lastpass in order to see how it behaves with the field, I discovered that they have a similar approach for this.

With password (note that they have a `value` attr inside the field) -
![lastpass1](https://user-images.githubusercontent.com/55033316/108070922-1af59800-708b-11eb-9961-c83200e5130b.png)

With text -
![lastpass2](https://user-images.githubusercontent.com/55033316/108070942-1fba4c00-708b-11eb-8755-92dbb123538c.png)

Signup page (with `autocompete=off` and no `value` attr)-
![lastpass3](https://user-images.githubusercontent.com/55033316/108071150-64de7e00-708b-11eb-8667-41c78fe6e17e.png)

The behaviour of lastpass is depicted in the following screenshot - 
![lastpass4](https://user-images.githubusercontent.com/55033316/108071278-8e97a500-708b-11eb-9c66-6df67e04725a.png)
The element masks the `eye` icon completely. However the functionality of both the icons remain independent of each other and work well due to different click positions. While the `eye` icon works on clicking anywhere over it, the lastpass element works only on clicking its right bottom.

ligmitz commented:
> A few things worth looking at:
> 
> * https://security.stackexchange.com/questions/185851/pitfalls-when-rolling-your-own-show-password-functionality
> * Several other things one finds searching for "show password feature security" on Google.
> 
> @ligmitz perhaps you can do a bit of research, reading those and posting your findings?

I went through a few of the articles and queries regarding the feature like https://webmasters.stackexchange.com/questions/100103/is-a-show-password-option-on-login-form-safe and https://security.stackexchange.com/questions/47557/why-is-there-a-show-password-functionality-in-some-login-forms which say that there is no security risk involved with it as such. The credentials are safe in any form unless the user have a malicious program or a plugin installed on there computer. It is advised to not keep it as a default behavior of-course. If a user feels safe enough to use the feature in their environment, the feature does not risk their data and credentials on its own.

timabbott commented:
First, thanks for doing all that research!  It's very helpful.  

We should plan to do this a bit more systematically with our password fields.  First, we have several places a user might enter a password:
```
$ git grep 'type="pass' static/templates/ templates/
static/templates/settings/account_settings.hbs:                                <input type="password" autocomplete="off" name="old_password" id="old_password" class="w-200 inline-block" value="" />
static/templates/settings/account_settings.hbs:                                <input type="password" autocomplete="off" name="new_password" id="new_password" class="w-200 inline-block" value=""
static/templates/settings/api_key_modal.hbs:                    <input type="password" autocomplete="off" name="password" id="get_api_key_password" value="" />
templates/zerver/login.html:                                <input id="id_password" name="password" class="required" type="password" autocomplete="off"
templates/zerver/register.html:                    <input id="ldap-password" class="required" type="password" name="password" required />
templates/zerver/register.html:                    <input id="id_password" class="required" type="password" name="password"
templates/zerver/reset_confirm.html:                    <input id="id_new_password1" class="required" type="password" name="new_password1"
templates/zerver/reset_confirm.html:                    <input id="id_new_password2" class="required" type="password" name="new_password2"
```

Second, should we be using `autocomplete=off` or the values like `autocomplete=new-password`?  
https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/autocomplete

I think given the number of call points, it's worth making this common JS code in `common.js` (which has our zxcvbn JS as well already), and making sure it works correctly for all of those contexts.  (E.g. we could probably do this for all inputs with `type="password")`.  We might also want to replace the IDs like `#password_eye_icon` with a more general CSS class name like `.password_visibility_toggle_on`, since we have 2 password fields on some pages, and duplicate IDs don't work.

ligmitz commented:
> Second, should we be using `autocomplete=off` or the values like `autocomplete=new-password`?

Looking at the docs, it seems that setting `autocomplete="off"` is preferable for fields where we enter the current password while `autocomplete="new-password"` is for fields where we confirm the password after entering it once such as resetting password.

Regarding the other `input` fields where we enter password, I have looked over some websites where they have this feature and they do not use `show password` feature for places where they reset password since confirming the password is used for verification and should not be visible I assume (though, I can update this if it is desired). For the other places where we have the `input` field, I will update this pull request to cover them all and shift the code to `common.js`.

ligmitz commented:
@timabbott I have made the updates.

Following is the appearance (all are functioning properly):

1. For api key - 

![api](https://user-images.githubusercontent.com/55033316/108382832-3b5e5780-722f-11eb-864f-b672bafd9288.png)

2. Changing password - 

![changepass1](https://user-images.githubusercontent.com/55033316/108382904-50d38180-722f-11eb-8c95-fc551fac7750.png) 
![changepass2](https://user-images.githubusercontent.com/55033316/108382927-55983580-722f-11eb-8eef-946a348ac341.png)

3. Login - 

![login](https://user-images.githubusercontent.com/55033316/108383026-6a74c900-722f-11eb-9743-6d132dd7f8b9.png)


4. Register - 

![register](https://user-images.githubusercontent.com/55033316/108383092-79f41200-722f-11eb-9f69-fd91c3080369.png)

I have not implemented the same for reset_confirm.html considering that it is a verification step and it might be insecure.
Rest all are working well with no side-effects.

ligmitz commented:
@timabbott The pull request is ready for review.

andersk commented:
I don’t see a reason for the reset page to be treated differently. If we are worried there’s something insecure about this, we shouldn’t do it at all; I think it’s fine though.

The sense of eye/eye-slash is reversed from other applications. The usual convention is that the icon indicates the current state, not the future state if you were to click it.

The Show API key dialog fails to toggle the eye/eye-slash icon at all.

In dialogs that can close an open again (change password, show API key), we need to reset the password fields to hidden. It might be prudent to reset all the password fields to hidden when the corresponding forms are submitted, since that might help some password managers recognize the password fields.

ligmitz commented:
> I don’t see a reason for the reset page to be treated differently. If we are worried there’s something insecure about this, we shouldn’t do it at all; I think it’s fine though.

Thankyou for the review. I have extended the pull request to cover the reset page as well.
> The sense of eye/eye-slash is reversed from other applications. The usual convention is that the icon indicates the current state, not the future state if you were to click it.

It's interesting that some websites use the current state format while others use the futute state (like lastpass above). Though I found after your comment that the current format is more common and hence made the change.
> The Show API key dialog fails to toggle the eye/eye-slash icon at all.

Sorry, it occured due to same class name within the page which altough changed the type of input but failed to toggle the icons. Fixed this !
> In dialogs that can close an open again (change password, show API key), we need to reset the password fields to hidden. It might be prudent to reset all the password fields to hidden when the corresponding forms are submitted, since that might help some password managers recognize the password fields.

For the modals part, I have added the functionality to revert back the icon and password type to default when the modal is closed or the submit buttons are clicked. For the resetting before submission in other parts, I tested the feature with different browsers and also lastpass (password manager) and found them responding the same way irrespective of the type of the input field. The approach they use for detecting these fields and inserting password may be different than what we expect and perhaps worth exploring.

andersk commented on static/js/portico/signup.js:
```diff
@@ -22,6 +22,11 @@ $(() => {
         });
     }
 
+    common.toggle_password("#ldap-password", ".password_visibility_toggle1");
+    common.toggle_password("#id_password", ".password_visibility_toggle2");
+    common.toggle_password("#id_new_password1", ".password_visibility_toggle_one");
+    common.toggle_password("#id_new_password2", ".password_visibility_toggle_two");
```
`password_visibility_toggle1` vs. `password_visibility_toggle_one` is confusing. The names should clearly indicate the difference. Or we could use a sibling (`#ldap-password ~ .password_visibility_toggle`) or adjacent sibling (`#ldap-password + .password_visibility_toggle`) selector and avoid coming up with separate names.

andersk commented on static/js/settings_account.js:
```diff
@@ -136,6 +136,27 @@ function update_user_custom_profile_fields(fields, method) {
+    $("#new_password").attr("type", "password");
+    const password_toggle_one = $(".password_visibility_toggle_one");
+    const password_toggle_two = $(".password_visibility_toggle_two");
+    if (password_toggle_one.hasClass("fa-eye")) {
+        password_toggle_one.toggleClass("fa-eye-slash fa-eye");
+    }
```
An unconditional `.removeClass("fa-eye").addClass("fa-eye-slash")` seems simpler.

andersk commented on static/js/settings_account.js:
```diff
@@ -370,6 +392,12 @@ export function set_up() {
             $(this).attr("href", settings_bots.encode_zuliprc_as_uri(data));
         });
+
+        $("#api_key_modal")
+            .find("[data-dismiss=modal]")
```
Or `$("#api_key_modal [data-dismiss=modal]")`.

andersk commented on static/js/settings_account.js:
```diff
@@ -416,6 +444,7 @@ export function set_up() {
         .find("[data-dismiss=modal]")
         .on("click", () => {
             clear_password_change();
+            reset_password_toggle_icons();
```
Should this just be part of the `clear_password_change` function?

ligmitz commented:
Thanks for the review @andersk. I have pushed the following updates to the pull request following your suggestions:
* Removed extra unnecessary classes to avoid confusion (used sibling selectors instead).
* Placed the code for `reset_password_toggle_icons` into `clear_password_change`.
* Placing `[data-dismiss=modal]` inside selector.
* Remove unnecessary conditional statements.

timabbott commented:
Is `aria-hidden` the right way to tag these?  Generally we use those for non-functional icons (E.g. ones that might sit next to an equivalent text label), and `aria-label` is preferred so that they're visible to screenreaders.  Also should these have a `title` attribute?

Worth checking what other products do on these questions.

timabbott commented:
Also, can you add a block comment in the `common.js` section around passwords (or a section added to the end of `docs/development/authentication.md` that we link to from a common, for better discoverability) listing the various policy decisions we have around authentication added in this PR -- (E.g. how to decide between `autocomplete=off` and `autocomplete=new-password`)?  Those will be a really helpful resource to help save time we'd otherwise spend revisiting this.

timabbott commented on static/js/settings_account.js:
```diff
@@ -383,6 +394,12 @@ export function set_up() {
+        $("#old_password").attr("type", "password");
+        $("#new_password").attr("type", "password");
+        const password_toggle_one = $("#old_password ~ .password_visibility_toggle");
+        const password_toggle_two = $("#new_password ~ .password_visibility_toggle");
```
These would be better named as E.g. `old_password_visibility_toggle`.  Also would it be better for this to call a helper exported from `common.js` twice rather than copying parts of the implementation here?

timabbott commented on static/js/common.js:
```diff
@@ -136,3 +136,17 @@ export function adjust_mac_shortcuts(key_elem_class, require_cmd_style) {
+
+        $(password_selector).toggleClass("fa-eye fa-eye-slash");
+    });
+}
```
I think most of this function could be extracted as a `toggle_password_visibility($password_field)`, and then have a wrapper function that also registers that as the click handler named something that'd be called as `common.setup_password_visibility_toggler("#ldap-password")`. -- that'd be a much nicer interface.

Also `password_id` is a confusing name for a jQuery object -- try something like `$password_field`.

timabbott commented on static/styles/portico/portico_signin.css:
```diff
@@ -521,7 +521,6 @@ html {
         display: none;
         vertical-align: top;
         position: relative;
-
```
Is this change intended to be part of this commit?

timabbott commented on static/styles/settings.css:
```diff
@@ -925,6 +925,28 @@ input[type="checkbox"] {
+            top: 30px;
+        }
+    }
+}
```
This is almost a duplicate of the previous block -- can it be deleted?  The only difference is 10px in the `left` attribute.

timabbott replied:
In an ideal world we'd just have a single copy of this in the shared `components.css`, if we can make the same CSS work for all the different fields.

timabbott commented on static/styles/zulip.css:
```diff
@@ -1090,6 +1090,16 @@ td.pointer {
+        top: 40px;
+    }
+}
+
```
Should this be in `settings.css`?  It also seems like a duplicate of the `settings.css` copy above.

ligmitz commented:
Addressed the following after previous review:
* Remove redundant styling.
* Splitting the main-function for handling the feature into two parts.
* Add documentation.
* Replaced `aria-hidden` with `aria-label`.
* Using a helper function to reset toggle icons.

timabbott commented on static/js/common.js:
```diff
@@ -131,3 +148,11 @@ export function adjust_mac_shortcuts(key_elem_class, require_cmd_style) {
+// for details on implementation.
+export function setup_password_visibility_toggle(password_field_id, password_selector) {
+    $("body").on("click", password_selector, () => {
+        toggle_password_visibility(password_field_id, password_selector);
```
Unless there's a good reason not to, click handlers that handle an event should use `e.preventDefault` and `e.stopPropagation` -- not doing so can result in unexpected behavior where a click does multiple things in a confusing way.

timabbott replied:
I pushed back with this change, since I'd made it manually before deciding this wasn't close enough for me to finish and merge today.  Also made a very small tweak to the `docs/` (changed 1-2 words) for better English readability.

ligmitz replied:
Okay.

ligmitz replied:
This is giving the error `Cannot read property preventDefault() of undefined` so the parameter passed in the click handler is undefined.

Signior-X replied:
`e` needs to be defined @ligmitz 
Like see you can do this way

![image](https://user-images.githubusercontent.com/56730716/112747621-274b0a00-8fd4-11eb-8f76-d24104f58173.png)

timabbott commented on templates/zerver/register.html:
```diff
@@ -148,21 +148,23 @@ <h1>{{ _('Create your account') }}</h1>
-                <div class="input-box">
-                    <input id="id_password" class="required" type="password" name="password"
+                <div class="input-box password-div">
+                    <input id="id_password" class="required" type="password" name="password" autocomplete="off"
```
Why isn't this `new-password` autocomplete?  I believe this is the registration form.

ligmitz replied:
Yes, should have been new-password. Will update this.

timabbott commented on templates/zerver/register.html:
```diff
@@ -148,21 +148,23 @@ <h1>{{ _('Create your account') }}</h1>
-                <div class="input-box">
-                    <input id="ldap-password" class="required" type="password" name="password" required />
+                <div class="input-box password-div">
+                    <input id="ldap-password" class="required" type="password" name="password" autocomplete="off" required />
```
I think this one is correct to be autocomplete=off, since we're asking for the current LDAP password, not a new password.  (In contrast with the next one in this file).

timabbott commented on templates/zerver/reset_confirm.html:
```diff
@@ -23,13 +23,14 @@ <h1>{{ _('Set a new password.') }}</h1>
                       value="{% if form.new_password1.value() %}{{ form.new_password1.value() }}{% endif %}"
                       maxlength="100"
                       data-min-length="{{password_min_length}}"
                       data-min-guesses="{{password_min_guesses}}" autofocus required />
```
I don't see an autocomplete=new-password here; is that intentional?

ligmitz replied:
I didn't add that property exclusively here since it wasn't present before but since we are amending the the fields, I'll add it here as well.

timabbott commented on templates/zerver/reset_confirm.html:
```diff
@@ -43,11 +44,12 @@ <h1>{{ _('Set a new password.') }}</h1>
                     <label for="id_new_password2" class="">{{ _('Confirm password') }}</label>
                     <input id="id_new_password2" class="required" type="password" name="new_password2"
                       value="{% if form.new_password2.value() %}{{ form.new_password2.value() }}{% endif %}"
                       maxlength="100" required />
```
This one should have autocomplete=off, right?

ligmitz replied:
Yep.

timabbott commented:
I posted a batch of comments from reading the code.  There's also a couple CSS changes this needs:
* There's a weird visual jump of the whitespace at least in "settings => change password" when toggling this element.  Please fix this and manually test every one of these widgets for similar issues.
* Hovering the icons should darken them; I'd copy styling from other icon-buttons we have (E.g. in compose), remembering to think about the night theme.

timabbott commented:
Also, one of the items discussed in how Google works is this:

> One of the additional thing they perform is changing the autocomplete attr from current-password to off on displaying the password.

What's your thinking on whether we should replicate this behavior?

ligmitz commented:
> There's a weird visual jump of the whitespace at least in "settings => change password" when toggling this element.  Please fix this and manually test every one of these widgets for similar issues.

I did see this but isn't sure about the reason. Will try to fix this.

ligmitz commented:
Made updates to the pull request:
1. Add and update `autocomplete` fields for input tags.
2. Fix the whitespace bug on toggling in `settings`.
3. Add styling for opacity to eye icons. (Since `i` tags are already handled in night mode so extra styling was not required for night mode)

Few points to note:
1. The `e.preventDefault()` gives error as `e` remains undefined in the click-handler.
2. On changing from day-mode to night-mode, the `settings` eye icon does not convert `type` password to `text` but this behaviour is resolved on reloading.
3. Regarding the change of `autocomplete` from `current-password` to `off`, we already keep this attribute to off for fields mostly (similar to dropbox).

ligmitz commented:
`On changing from day-mode to night-mode using the /night zcommand without reloading, the settings eye icon does not convert type password to text but this behaviour is resolved on reloading.` Fixed this by replacing
```code
    $("body").on("click", password_selector, () => {
        toggle_password_visibility(password_field_id, password_selector);
    });
```
with

```code
    $(password_selector).on("click", () => {
        toggle_password_visibility(password_field_id, password_selector);
    });
```
Apparently, since night-mode adds and removes class from body tag, it behaved incorrectly with the former code block.

Signior-X commented:
Like I was reviewing this PR. The first thing I noticed is this.
![image](https://user-images.githubusercontent.com/56730716/112747305-244f1a00-8fd2-11eb-8f2d-9b16b8fb505d.png)

Why these two different ones are showing?
Also, these two have the same functions
![image](https://user-images.githubusercontent.com/56730716/112747321-4b0d5080-8fd2-11eb-9dd1-d5f8e7edc36d.png)

Any idea related to this?

Signior-X reviewed (commented):
Commented here what needs to be changed. Rest the PR works fine on my machine.

ligmitz commented:
```quote
`e` needs to be defined @ligmitz
Like see you can do this way
```
@Signior-X The parameter is passed in the click-handler but this only declares the variable and extends its scope, but the error is not `e is not defined`, but instead it is `Cannot read property preventDefault of undefined`.

Signior-X commented:
For me, it is not coming undefined. And it can not be undefined, every click listener gives some event.
![image](https://user-images.githubusercontent.com/56730716/112753307-b6681a00-8ff4-11eb-9723-0591cf13937c.png)

This is the code so that you can cross-check.
![image](https://user-images.githubusercontent.com/56730716/112753361-eadbd600-8ff4-11eb-9502-d1db088510e2.png)

ligmitz commented:
@Signior-X Thanks for the help. The event does exist and it was showing `undefined` in the node-test which led me into believing it does not exist. I have updated the tests now with the event properties so that its all working well now.

timabbott commented:
> Apparently, since night-mode adds and removes class from body tag, it behaved incorrectly with the former code block.

(From https://github.com/zulip/zulip/pull/17305#issuecomment-806053013).  This explanation doesn't make sense to me -- could this be this is probably a similar node testing issue similar to the "e is undefined" bug?  `$(body)` definitely works in both contexts -- you can test in the JS console.

timabbott commented:
Since we're doing another round trip on this, can you also test the CSS with browser zoom to make sure the icon still is centered nicely with that?  The `top/right` usage seems potentially risky to me from the standpoint; I'd have expected `vertical-align: middle` or similar to be more reliable for our goals here.

One other thought -- I think this PR would have been better done as a series of commits, with the first implementing the common framework and changing one form, and then each additional commit changing an additional form.  That would allow me to merge individual ones incrementally once we've finished testing them, even if other forms have issues (like the whitespace one we had before).  

It might be a good exercise for you to split this commit using `git rebase -i`, `git checkout <commit_id> path` and `git add -p` and the like to split this into that structure.  So if you have a bit of time to spend doing that, it's probably worth doing so.  (I'd expect it to take <10 minutes if you know the right Git techniques enough, but almost nobody does).

ligmitz commented:
> Since we're doing another round trip on this, can you also test the CSS with browser zoom to make sure the icon still is centered nicely with that? The `top/right` usage seems potentially risky to me from the standpoint; I'd have expected `vertical-align: middle` or similar to be more reliable for our goals here.

I have tried to keep all the elements in the center and have attached the close view screenshots as well. The `vertical-align: middle` property isn't helpful here as we use `position: absolute` as well as due to different heights of divisions and responsiveness. I checked how other websites handle css and they use similar properties.

1. For register
![register_pass](https://user-images.githubusercontent.com/55033316/113557203-458bb800-961b-11eb-84f3-e914620f0cba.png)

2. For reset password
![reset_pass](https://user-images.githubusercontent.com/55033316/113557212-49b7d580-961b-11eb-96d2-6d436ca089db.png)

3. For login password
![login_pass](https://user-images.githubusercontent.com/55033316/113557222-4cb2c600-961b-11eb-9447-c70fffad8ee9.png)

4. for reset password in settings 
![settings_pass](https://user-images.githubusercontent.com/55033316/113557234-51777a00-961b-11eb-9279-acdbf249e5ed.png)

5. For api key in settings
![pass_api](https://user-images.githubusercontent.com/55033316/113557251-55a39780-961b-11eb-8db5-fd61f7a5a5b1.png)


> It might be a good exercise for you to split this commit using `git rebase -i`, `git checkout <commit_id> path` and `git add -p` and the like to split this into that structure. So if you have a bit of time to spend doing that, it's probably worth doing so. (I'd expect it to take <10 minutes if you know the right Git techniques enough, but almost nobody does).

It took me only 5 minutes as I used `git reset` :)

ligmitz commented:
@timabbott I have split the commit into 4 individual ones. This is ready for a final review.

ligmitz commented:
> > Apparently, since night-mode adds and removes class from body tag, it behaved incorrectly with the former code block.
> 
> (From [#17305 (comment)](https://github.com/zulip/zulip/pull/17305#issuecomment-806053013)). This explanation doesn't make sense to me -- could this be this is probably a similar node testing issue similar to the "e is undefined" bug? `$(body)` definitely works in both contexts -- you can test in the JS console.

This is something I found while manually testing and not as part of the node-tests. `$(body)` definitely worked fine but just had the bug while shifting from one theme to another (using zcommand), without refreshing the page before testing again.

timabbott commented on templates/zerver/login.html:
```diff
@@ -67,11 +67,12 @@ <h1 class="get-started">{{ _("Log in to Zulip") }}</h1>
                                   {% if email %} autofocus {% endif %}
                                   required />
                                 <label for="id_password" class="control-label">{{ _('Password') }}</label>
+                                <i class="fa fa-eye-slash password_visibility_toggle" role="button" aria-label="{{ _('Toggle password visibility') }}"></i>
```
I think we need to set `title` here, or perhaps setup TippyJS -- otherwise hovering the element does not display the label.  @amanagr FYI in case you have advice.

timabbott commented on static/js/common.js:
```diff
@@ -2,6 +2,23 @@ import $ from "jquery";
+        password_field.attr("type", "password");
+    }
+
+    $(password_selector).toggleClass("fa-eye fa-eye-slash");
```
I think this would be better written as an if/else block rather than doing toggles.  Or probably even better, we should just make this two functions, one for `fa-eye` and the other for clicking on a `fa-eye-slash`.

As it is, the code is just kinda hard to read.

timabbott commented on static/js/common.js:
```diff
@@ -2,6 +2,23 @@ import $ from "jquery";
+export function reset_password_toggle_icons(password_field, password_selector) {
+    $(password_field).attr("type", "password");
+    $(password_selector).removeClass("fa-eye").addClass("fa-eye-slash");
+}
```
This first commit doesn't use this function.

ligmitz commented:
Updated the pull request to give `title` to fields in `portico` and `tippy tooltip` to fields in settings modal.

timabbott commented:
@ligmitz OK, I did another review pass on this and added a few commits of my own that clean up the implementation and adjust the choice of strings to match what other companies do.

The main thing we still need here is that my Tippy implementation doesn't work -- it always shows "Show password".  Can you debug that and add a new commit fixing it (possibly one meant to be squashed into my last commit), and then also do another round of review / manual testing on my changes?  Otherwise I think this is ready to merge.

ligmitz commented:
Yeah, Sure will do.

amanagr commented on static/js/common.js:
```diff
@@ -134,3 +134,48 @@ export function adjust_mac_shortcuts(key_elem_class, require_cmd_style) {
+}
+
+export function setup_password_visibility_toggle(password_field_id, password_selector, opts = {}) {
+    opts = {...opts, tippy_tooltips: false};
```
You should open `...opts` after setting defaults.

amanagr commented on static/js/common.js:
```diff
@@ -134,3 +135,51 @@ export function adjust_mac_shortcuts(key_elem_class, require_cmd_style) {
+function set_password_toggle_label(password_selector, label, tippy_tooltips) {
+    $(password_selector).attr("aria-label", label);
+    if (tippy_tooltips) {
+        const tippy_selectors = tippy(password_selector);
```
`tippy_instances` is better variable name as it is commonly used in the tippy docs. Also, use `instance` below.

andersk commented on frontend_tests/node_tests/common.js:
```diff
@@ -10,6 +10,11 @@ mock_cjs("jquery", $);
 
 const noop = () => {};
 
+mock_cjs("tippy.js", (selector) => {
+    $(selector)[0]._tippy = noop;
+    $(selector)[0]._tippy.setContent = noop;
+});
```
`tippy.js` is actually an ES module (see `"module": "dist/tippy.esm.js"` in `node_modules/tippy.js/package.json`) that we default-import, so it should be mocked with `mock_esm`:

```suggestion
mock_esm("tippy.js", {
    default: (selector) => {
        $(selector)[0]._tippy = noop;
        $(selector)[0]._tippy.setContent = noop;
    },
});
```

ligmitz replied:
Ah yes, will update.

ligmitz replied:
Updated.

timabbott commented:
Code lgtm; I'll merge once CI passes.

timabbott commented:
Merged, thanks @ligmitz!
